app/validators/questionnaire_schema.py

```diff
@@ -0,0 +1,298 @@
+
+                block_path = path_list[: path_list.index("blocks") + 2]
+
+                string_path = ".".join(block_path)
+                # Since unique_ids_per_block is a set, duplicate ids will only be recorded once within the block.
+                unique_ids_per_block[string_path].add(value)
+            else:
+                non_block_ids.append(value)
+
+        for block_ids in unique_ids_per_block.values():
+            all_ids.extend(block_ids)
+        all_ids.extend(non_block_ids)
+
+        return all_ids
+
+    @cached_property
+    def id_paths(self):
+        """
+        These values will be returned with the json path to them through the object e.g.
+            - 'sections.[0].groups[0].blocks[1].question_variants[0].question.question-2'
+
+        Returns: generator yielding (path, value) tuples
+        """
+        ignored_paths = [
+            "routing_rules",
+            "skip_conditions",
+            "when",
+            "edit_block.question.answers",
+            "add_block.question.answers",
+            "remove_block.question.answers",
+            "edit_block.question_variants.answers",
+            "add_block.question_variants.answers",
+            "remove_block.question_variants.answers",
+        ]
+
+        for match in parse("$..id").find(self.schema):
+            full_path = str(match.full_path)
+            if not any(ignored_path in full_path for ignored_path in ignored_paths):
+                yield str(match.full_path)[:-3], match.value
+
+    @cached_property
+    def answer_id_to_option_values_map(self):
+        answer_id_to_option_values_map = defaultdict(set)
+
+        for answer in self.answers:
+            if "options" not in answer:
+                continue
+
+            answer_id = answer["id"]
+            option_values = [option["value"] for option in answer["options"]]
+
+            answer_id_to_option_values_map[answer_id].update(option_values)
+
+        return answer_id_to_option_values_map
+
+    @cached_property
+    def answers(self):
+        for question, _ in self.questions_with_context:
+            for answer in question["answers"]:
+                yield answer
+
+    @lru_cache
+    def get_answer(self, answer_id):
+        return self.answers_with_context[answer_id]["answer"]
+
+    @lru_cache
+    def get_group(self, group_id):
+        return self.groups_by_id[group_id]
+
+    @lru_cache
+    def get_section(self, section_id):
+        return self.sections_by_id[section_id]
+
+    @lru_cache
+    def get_block(self, block_id):
+        return self.blocks_by_id[block_id]
+
+    @lru_cache
+    def get_blocks(self, **filters):
+        conditions = []
+        for key, value in filters.items():
+            conditions.append(f'@.{key}=="{value}"')
+
+        if conditions:
+            final_condition = " & ".join(conditions)
+            return jp.match(f"$..blocks[?({final_condition})]", self.schema)
+        return self.blocks
```

Comment:
Originally, there were a number of other methods calling this block method with different filters, hence why it currently is fairly generic. It sits alongside the get_other_blocks which returns the opposite set of blocks which don't match the conditions which is called a number of times. These replace a number of methods with overly verbose names.

I kind of disagree on the calling method constructing the query. I don't think it should be necessary to understand how to use jsonpath in order to query our schemas. I also think the logic for schema navigation should be kept in a single class.

Reply:
```
self.get_blocks(type="ListCollectorDrivingQuestion", for_list=list_name)

f"$..blocks[?(@.type=="ListCollectorDrivingQuestion" & @.for_list=={list_name}")]"
```

Reply:
Hmm, ok, annoyingly this seems more to do with the complex nature of JSONPath queries. We should keep an eye on this so that we don't end up with a proliferation of very similar methods e.g. the same as this but for groups or sections.

Reply:
Yeah, I'd prefer to have one or two flexible generic methods rather than a whole bunch of methods for single calls.